openevse: Expand the documentation. (#45822)
home-assistant/home-assistant.github.io: source/_integrations/openevse.markdown

Use the integration to monitor your <abbr title="electric vehicle">EV</abbr> charging sessions, track energy consumption for the [Energy dashboard](/home-energy-management), adjust the charging current, and react to charger state changes in automations.

## Use cases

The OpenEVSE integration is useful in scenarios like the following:

- Track how much energy each charging session uses and feed it into the [Energy dashboard](/home-energy-management) to see your home’s EV charging costs over time.
- Lower the charge rate during peak electricity hours and raise it again when off-peak rates start.
- Match the charge rate to the surplus power your solar panels produce, so your car charges from clean energy whenever possible.
- Get a notification when your car is plugged in, when charging starts or finishes, or when the charger reports a fault.
- Pause charging when other large loads (such as an oven or heat pump) are running, to stay below the limit of your electrical service.

## Supported devices

> This integration supports OpenEVSE charging stations that run the official [OpenEVSE Wi-Fi firmware](https://github.com/OpenEVSE/openevse_esp32_firmware) on either an ESP8266 or an ESP32 Wi-Fi module. The Wi-Fi module is what exposes the local HTTP API and websocket interface that this integration uses.

Reviewer: ```suggestion
This integration supports OpenEVSE charging stations that run the official [OpenEVSE Wi-Fi firmware](https://github.com/OpenEVSE/openevse_esp32_firmware) on either an ESP8266 or an ESP32 Wi-Fi module. The Wi-Fi module is what exposes the local HTTP API and WebSocket interface that this integration uses.
```

{% configuration_basic %}
Host:
  description: "The IP address or hostname of your OpenEVSE charger on your local network, for example `192.168.1.50` or `openevse.local`. This is the same address you use to open the OpenEVSE web interface in a browser."
Username:
  description: "The username for the OpenEVSE web interface. Only required if you have enabled authentication on the charger. Leave empty if you have not set a username."
Password:
  description: "The password for the OpenEVSE web interface. Only required if you have enabled authentication on the charger. Leave empty if you have not set a password."
{% endconfiguration_basic %}

If the credentials on the charger change later, Home Assistant prompts you to re-enter them through the reauthentication flow without removing and re-adding the integration.

## Supported functionality

The integration adds a single device per charger to Home Assistant and exposes the following entities. Some entities are disabled by default to keep the device page tidy. You can enable them from the device page if you need them.

### Sensors

> 

Reviewer: In the strings file, I also see the following sensors. Would it be worth adding those to the list?
- Mode
- OTA update
- Available current
- Smoothed available current
- Charge rate

Author: Good thought, actually extra unused translations. I'll get them deleted.

- **Uptime** and **Free memory**: Information about the Wi-Fi module itself.

### Binary sensors

- **Vehicle connected**: On when a vehicle is plugged in to the charger.
- **Divert active**: On when the charger is in solar divert (eco) mode and is matching its output to available surplus power.
- **Shaper active**: On when the power shaper is actively limiting the charge rate to stay below a household power limit.
- **Limit active** (disabled by default): On when an energy, time, or session limit is currently being enforced.
- **Ethernet connected** (disabled by default): On when the charger is using its wired Ethernet connection instead of Wi-Fi.
- **MQTT connected** (disabled by default): On when the charger has an active MQTT connection.

### Number

- **Charge rate**: Sets the maximum current the charger will deliver, in amperes. The minimum and maximum follow the limits that the charger reports for itself, so you cannot set a value outside the safe range for your hardware.

## Examples

### Notify when your car is plugged in

This automation sends a notification whenever a vehicle is connected to the charger.

> ```yaml

Reviewer: ```suggestion
{% example %}
```

> automations:

Reviewer: ```suggestion
automation:
```

  - alias: "Notify when EV is plugged in"
    triggers:
      - trigger: state
        entity_id: binary_sensor.openevse_vehicle_connected
        from: "off"
        to: "on"
    actions:
      - action: notify.mobile_app_my_phone
        data:
          title: "EV plugged in"
          message: "Your car is now connected to the charger."

> ```

Reviewer: ```suggestion
{% endexample %}
```

### Lower the charge rate during peak hours

This automation reduces the charge rate to 10A between 5 PM and 9 PM, then returns it to 32A afterwards. Adjust the times and current values to match your electricity tariff and your charger’s limits.

> ```yaml

Reviewer: ```suggestion
{% example %}
```
We now have a liquid tag to wrap YAML examples. 
https://github.com/home-assistant/home-assistant.io/blob/current/source/_integrations/_integration_docs_template.markdown?plain=1#L175

> automations:

Reviewer: ```suggestion
automation:
```

  - alias: "OpenEVSE peak rate"
    triggers:
      - trigger: time
        at: "17:00:00"
    actions:
      - action: number.set_value
        target:
          entity_id: number.openevse_charge_rate
        data:
          value: 10

  - alias: "OpenEVSE off-peak rate"
    triggers:
      - trigger: time
        at: "21:00:00"
    actions:
      - action: number.set_value
        target:
          entity_id: number.openevse_charge_rate
        data:
          value: 32

> ```

Reviewer: ```suggestion
{% endexample %}
```

### Add EV charging to the Energy dashboard

Use the **Total energy usage** sensor as an individual device on the [Energy dashboard](/home-energy-management):

1. Go to {% my energy title="**Settings** > **Dashboards** > **Energy**" %}.
2. Under **Individual devices**, select **Add device**.
3. Choose the **Total energy usage** sensor from your OpenEVSE charger.

The dashboard then shows how much energy your car consumes compared to the rest of your home.

## Data updates

> The OpenEVSE integration uses a {% term "push" %} model. After it connects, it opens a websocket to the charger, and the charger sends new data as soon as a value changes. This means entity states reflect what the charger is doing within seconds, without waiting for a polling interval.

Reviewer: ```suggestion
The OpenEVSE integration uses a {% term "push" %} model. After it connects, it opens a WebSocket to the charger, and the charger sends new data as soon as a value changes. This means entity states reflect what the charger is doing within seconds, without waiting for a polling interval.
```

> As a safety net, the integration also polls the charger every five minutes. This catches the rare case where the websocket connection drops without a clean disconnect, so the entities are refreshed on the next poll even if the push channel is interrupted.

Reviewer: ```suggestion
As a safety net, the integration also polls the charger every five minutes. This catches the rare case where the WebSocket connection drops without a clean disconnect, so the entities are refreshed on the next {% term polling poll %} even if the push channel is interrupted.
```

4. Confirm that no firewall on your network is blocking traffic between Home Assistant and the charger.

### Setup fails with "Invalid authentication"

The charger has authentication enabled and the username or password you entered is incorrect. Re-enter the credentials you use to log in to the OpenEVSE web interface. If you have forgotten them, reset the credentials from the charger’s web interface.

### Home Assistant asks you to reauthenticate

If you change the username or password on the charger, Home Assistant detects the failed authentication and starts a reauthentication flow. Select **Reconfigure** on the integration’s notification and enter the new credentials. The integration reconnects without losing any of your existing entities or automations.

### Entities show "Unavailable"

If most entities go unavailable at the same time, the integration has lost contact with the charger. Check the following:

1. The charger has power and its status LED is on.
2. The charger’s web interface still loads in a browser.
3. The charger has not picked up a new IP address from your router. Use a static IP or a DHCP reservation to avoid this.

> After the charger is reachable again, entities recover automatically on the next websocket message or polling cycle.

Reviewer: ```suggestion
After the charger is reachable again, entities recover automatically on the next WebSocket message or {% term polling %} cycle.
```

- Pause charging when other large loads (such as an oven or heat pump) are running, to stay below the limit of your electrical service.

## Supported devices

This integration supports OpenEVSE charging stations that run the official [OpenEVSE Wi-Fi firmware](https://github.com/OpenEVSE/openevse_esp32_firmware) on either an ESP8266 or an ESP32 Wi-Fi module. The Wi-Fi module is what exposes the local HTTP API and WebSocket interface that this integration uses.

The OpenEVSE Advanced Series chargers are known to work with this integration. If your charger runs the OpenEVSE Wi-Fi firmware and exposes the local web interface in a browser, it is expected to work with this integration.

### Unsupported devices

The following devices are not supported:

- OpenEVSE chargers without the Wi-Fi module, since they have no network interface to communicate with.
- Older OpenEVSE Wi-Fi firmware versions that do not expose the modern HTTP API. If the local web interface does not show energy totals or a status page, update the firmware first.

## Prerequisites

- The OpenEVSE charger is on the same network as Home Assistant.

{% include integrations/config_flow.md %}

> In most cases the charger is discovered automatically through zeroconf and shown as a new discovered device. If automatic discovery does not find your charger, you can add it manually by entering its IP address or hostname.

Reviewer: ```suggestion
In most cases, the charger is discovered automatically and shown as a newly discovered device. If automatic discovery does not find your charger, you can add it manually by entering its IP address or hostname.
```
